hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
     }
-    return 0;
+    return SYSTEM_ERROR_NONE;
 }
 
-} // ::
+int BleObject::Broadcaster::suspend(void) {
+    if (isAdvertising_) {
+        int ret = ble_gap_stop_advertising();
+        CHECK_RETURN(ret, "ble_gap_stop_advertising");
+        advPending_ = true;
+    }
+    return SYSTEM_ERROR_NONE;
+}
 
-int ble_init(void* reserved) {
-    if (!nrf_sdh_is_enabled()) {
-        LOG(ERROR, "SoftDevice is not enabled");
-        return BLE_ERROR_INVALID_STATE;
+int BleObject::Broadcaster::resume(void) {
+    if (advPending_) {
+        int ret = startAdvertising();
+        CHECK_RETURN(ret, nullptr);
+        advPending_ = false;
     }
-    // Configure the stack using the SDK's default settings
-    uint32_t ramStart = 0; // Start address of the application RAM
-    ret_code_t ret = nrf_sdh_ble_default_cfg_set(CONN_CFG_TAG, &ramStart);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "nrf_sdh_ble_default_cfg_set() failed: %u", (unsigned)ret);
-        return halError(ret);
+    return SYSTEM_ERROR_NONE;
+}
+
+void BleObject::Broadcaster::toPlatformAdvParams(ble_gap_adv_params_t* params, const hal_ble_adv_params_t* halParams) {
+    memset(params, 0x00, sizeof(ble_gap_adv_params_t));
+    params->properties.type = halParams->type;
+    params->properties.include_tx_power = false; // FIXME: for extended advertising packet
+    params->p_peer_addr = nullptr;
+    params->interval = halParams->interval;
+    params->duration = halParams->timeout;
+    params->filter_policy = halParams->filter_policy;
+    params->primary_phy = BLE_GAP_PHY_1MBPS;
+}
+
+void BleObject::Broadcaster::toPlatformAdvData(ble_gap_adv_data_t* data) {
+    data->adv_data.p_data = advData_;
+    data->adv_data.len = advDataLen_;
+    data->scan_rsp_data.p_data = scanRespData_;
+    data->scan_rsp_data.len = scanRespDataLen_;
+}
+
+int BleObject::Broadcaster::configure(const hal_ble_adv_params_t* params) {
+    int ret;
+    ble_gap_adv_data_t bleGapAdvData;
+    toPlatformAdvData(&bleGapAdvData);
+    if (params == nullptr) {
+        ret = sd_ble_gap_adv_set_configure(&advHandle_, &bleGapAdvData, nullptr);
+    } else {
+        ble_gap_adv_params_t bleGapAdvParams;
+        toPlatformAdvParams(&bleGapAdvParams, params);
+        LOG_DEBUG(TRACE, "BLE advertising interval: %.3fms, timeout: %dms.",
+                  bleGapAdvParams.interval*0.625, bleGapAdvParams.duration*10);
+        ret = sd_ble_gap_adv_set_configure(&advHandle_, &bleGapAdvData, &bleGapAdvParams);
     }
-    LOG_DEBUG(TRACE, "RAM start: 0x%08x", (unsigned)ramStart);
-    // Enable the stack
-    ret = nrf_sdh_ble_enable(&ramStart);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "nrf_sdh_ble_enable() failed: %u", (unsigned)ret);
-        return halError(ret);
+    NRF_CHECK_RETURN(ret, "sd_ble_gap_adv_set_configure");
+    return SYSTEM_ERROR_NONE;
+}
+
+int8_t BleObject::Broadcaster::roundTxPower(int8_t value) {
+    uint8_t i = 0;
+    for (i = 0; i < sizeof(validTxPower_); i++) {
+     
```

Comment:
Should we panic here?

Reply:
>BleObject::init()
>        if (sdRamEnd > appRamStart) {
            LOG(ERROR, "Need to change APP_RAM_BASE in linker script to be large than: 0x%08x", (unsigned)sdRamEnd - 0x20000000);
        }

I'd like using `LOG` for easy debugging.

Reply:
The device will end up in an invalid state if SD RAM overlaps system-part1 RAM, so I'd argue that we definitely need to panic here to indicate an issue, because normal execution cannot continue after this point.